### PR TITLE
Fix partial fills

### PR DESCRIPTION
**Pull request by fleupold**
DEX solvers are supposed to attempt filling partially fillable orders completely at first, but if unsuccessful, binary search their way to a smaller appropriate fill size (cf. logic in `domain::solver::dex::fills`).

Currently fill size is reduced when the API returned a `NotFound` error. While this may indicate a liquidity issue (for extremely large orders) it doesn't cover the case where liquidity exists but not for the user specified price.

This PR adds a check to the `try_solve` method that inspects the proposed swap and - if it doesn't satisfies the user's limit price - also reduces the fill size for the next loop.

This should allow DEX solvers to be more competitive for partial fills.

### Test Plan

Adjusted the unit test to show the behaviour

### CODE REVIEW

**Comment by sunce86** [comment]
AFAIS we retry smaller amounts in the following auctions right? 

Do you think we should have a follow up PR to execute the loop within the same auction?

**Comment by fleupold** [comment]
I think given the size of the orderbook and our rate limits it's fine to wait for a few blocks/auctions to find the right fill (I'd rather try more orders in a single auciton than more fill sizes)

**Review by sunce86 — APPROVED** [review]

**Review by squadgazzz — APPROVED** [review]

File: src/domain/dex/mod.rs [hunk 1]
```diff
@@ -147,6 +147,11 @@ impl Swap {
         }
         .into_solution(gas_price, sell_token, score)
     }
+
+    pub fn satisfies(&self, order: &domain::order::Order) -> bool {
+        self.output.amount * (order.sell.amount + order.fee.0)
+            >= self.input.amount * order.buy.amount
```
Comment (MartinquaXD) [inline src/domain/dex/mod.rs:153]: Doesn't this need to take into account that the meaning of `self.output` and `self.input` can be switched depending on `sell` or `buy` orders?
If I'm not mistaken this formular is currently only correct for `sell` orders, no?

Comment (fleupold) [reply src/domain/dex/mod.rs]: Monstercatch!



Adjusted the existing test to catch this.

Comment (fleupold) [reply src/domain/dex/mod.rs]: Actually, never mind (buy order logic was already covered before). The in and out amount are always from the pov of the DEX aggregator and so they don't depend on the user order type.

**Review by MartinquaXD — APPROVED** [review]

**Pull request merged**